### ISSUE TITLE
Edge devtools extension doesn't work on remote hosts.

### ISSUE THREAD

**Issue by aaronpowell** [microsoft/vscode-js-debug#1024]
**Describe the bug**
I'm using the [Edge DevTools Extension](https://github.com/microsoft/vscode-edge-devtools) to inspect the DOM/network from within VS Code. Using the new inspector integration, as per [this comment](https://github.com/microsoft/vscode-edge-devtools/issues/294#issuecomment-855569512) from @mliao95, doesn't work in a remote environment (WSL or containers).

When clicking the inspector icon, rather than launching the inspector it triggers prompt to install the extension, even though it's already installed.

This is most likely because [this check](https://github.com/microsoft/vscode-js-debug/commit/0ced734b13bfa3f7c0226a407c0409a7166b2ee5#diff-c2f50c66eca36fd81ef178ec45bd939085c9d361c29fbdd01b87b883b6a9799cR35) fails in a remote instance [as `vscode.extensions.all` doesn't work as expected in remotes](https://github.com/microsoft/vscode/issues/82633).

**To Reproduce**
1. Open a remote workspace (WSL or containers)
1. Install the nightly js-debug extension
2. Install the Edge devtools extension
3. Launch a browser with client-side debugging (either Open Link or a launch task of `pwa-edge`)
4. Click the inspector icon
5. **Expected** Edge extension to show the inspector. **Actual** Prompted to install the extension

**VS Code Version:**
```
Version: 1.57.0-insider (system setup)
Commit: bf84ee619e0b67443d0307c04348fa575bc3c363
Date: 2021-06-07T12:46:22.208Z
Electron: 12.0.9
Chrome: 89.0.4389.128
Node.js: 14.16.0
V8: 8.9.255.25-electron.0
OS: Windows_NT x64 10.0.19043
```

**Additional context**

I've cloned the `js-debug` extension, removed the check for the Edge extension and side-loaded it into a container. That does result in the expected behaviour of _click inspector -> get Edge inspector_ but the Edge inspector doesn't properly connect. The DOM inspector doesn't show any DOM nodes and the Network tab doesn't have any traffic in it. My assumption is that the debugger isn't forwarding requests through properly in a remote environment, but how that all works is beyond what I've been able to discover.

Here's a video of it in action, using my modified version of the `js-debug` extension. An issue has also been opened on the [vscode-edge-devtools issue register](https://github.com/microsoft/vscode-edge-devtools/issues/413), to keep that team in the loop too.

https://user-images.githubusercontent.com/434140/121441188-aabdaa00-c9cc-11eb-9322-cd43fe8165ad.mp4

**Comment by connor4312** [comment]
Two things that need to happen:

1. Aaron mentioned in DM that vscode.extensions.all might not work correctly with remotes. This may require a VS Code core change to fix.
2. We need to forward the proxy port to the UI side. This needs to happen in js-debug since `vscode.workspace.openTunnel` is still in proposed. Perhaps some new `forwardRemote: true` option in the command, that will cause js-debug to automatically forward when running in a remote.

_or_ the devtools extension could become a workspace extension. This may have other complications depending on its setup.

**Comment by aaronpowell** [comment]
Yes, the way `vscode.extensions.all` works predates the notion of remotes so it doesn't work the way you'd expect it to, more info can be found here: https://github.com/microsoft/vscode/issues/82633 and I think it returns the _workspace_ extensions, and since the Edge one is a UI extension, that'd be why it doesn't show up.

**Comment by bgoddar** [comment]
> or the devtools extension could become a workspace extension. This may have other complications depending on its setup.

Interesting. I am not familiar with how workplace extensions work, I'm happy to look into documentation to see what this would require on our end if we determine that would be best. Let me know if you are unable to come up with a fix in the vscode/js-debug side of things

**Comment by connor4312** [comment]
Here's the whole shebang on remote extensions, to specify a workspace extension you would use the `extensionKind`: https://code.visualstudio.com/api/advanced-topics/remote-extensions, which is [currently UI](https://github.com/microsoft/vscode-edge-devtools/blob/master/package.json#L10).

However, if that was changed outright for the dev tools extension then launching browsers running in a remote would become difficult unless you solely depend on js-debug for doing so.

**Comment by aaronpowell** [comment]
Using a workspace extension is an interesting idea, but I do see there being complications.

There's two ways to debug Edge with the extension, attach to a browser instance that is managed by the `js-debug` extension (which the inspector icon does) or to launch a remotely controlled browser (which uses puppeteer to launch).

The first scenario might be entirely possible in a workspace extension, as it's `js-debug` that has launched the browser (or at least, that's my understanding of https://github.com/microsoft/vscode-js-debug/blob/main/src/ui/debugLinkUI.ts#L62-L88).

But the second scenario is likely to be the problem. As I mentioned, it uses puppeteer to launch (https://github.com/microsoft/vscode-edge-devtools/blob/master/src/utils.ts#L352-L368) and it receives the path-on-disk to the Edge binary. If puppeteer is running within the remote, it won't have access to the host to launch the binary, it'd want to launch one within the remote. WSL it's _technically_ possible, since WSL can launch a Windows binary from Linux, containers (and SSH) pose a problem. I'm not sure if a workspace extension can signal the VS Code host to execute something there.

**Comment by aaronpowell** [comment]
@bgoddar if you're missing any context or anything, ping me on Teams and we can jump on a call. 😃

**Comment by connor4312** [comment]
js-debug uses a [companion UI extension](https://github.com/microsoft/vscode-js-debug-companion) to launch the browser, as it too is a workspace extension. However I don't know whether I want the launch process to become contract.

**Comment by aaronpowell** [comment]
Oh so that's what the companion extension does. I had been wondering (but hadn't waded into its source yet)

**Comment by connor4312** [comment]
1. Check to see if the edge debug extension is running in the workspace ([e.g.]https://github.com/microsoft/vscode-js-debug/blob/baacf1949db03a34ee8ad2160cc77f21c371a77a/src/extension.ts#L37))
2. If not, pass `forwardToUi: true` to `extension.js-debug.requestCDPProxy`
3. I'll return back a forwarded port

**Comment by aaronpowell** [comment]
Just tested this on the codebase I used in the video above and I'm not able to get the "connect Edge" button to appear (using the stable release of the Edge extension). Were you able to verify it working @connor4312?

**Comment by connor4312** [comment]
I think @bgoddar still needs to adopt+publish the changes on the Edge devtools extension side of things.

**Comment by aaronpowell** [comment]
I'll keep an eye on that

**Comment by bgoddar** [comment]
> I think @bgoddar still needs to adopt+publish the changes on the Edge devtools extension side of things.

Correct. I'm hoping to have this fixed published early this week

**Comment by bgoddar** [comment]
Just published with version 1.2.1. Please try it out and let us know what you think!

**Comment by aaronpowell** [comment]
@bgoddar just tested on the workspace I've been using for testing and still doesn't look like it works.

```
Version: 1.60.0-insider (system setup)
Commit: 1ee61f368ee0570feeb220605578a8768d99e762
Date: 2021-08-09T06:11:00.971Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.22000
```

JS Debugger Nightly v2021.8.717
Microsoft Edge Tools for VS Code v1.2.1

ping me on Teams if you want a demo of what I'm seeing.

**Comment by connor4312** [comment]
I can confirm this seems to work for me using those same versions over remote SSH. To confirm you're on the latest version of both, make sure that the port is getting forwarded in the "Ports" panel:

![](https://memes.peet.io/img/21-08-9e129b15-0ffc-4054-9afa-e8dd5e0d6966.png)

You could also `Toggle Developer Tools` and see if you notice any errors in the console

**Comment by aaronpowell** [comment]
_Testing using Remote Containers extension._

I see the `Browser Debug Tunnel` port open up, but I get the following error in the VS Code devtools when I use `Debug: Open Link`.

![image](https://user-images.githubusercontent.com/434140/128952339-c6d1700d-381e-4e3b-8441-82b95fd90d91.png)

Tried debugging backwards through the callstack and ha, I'm soooo lost!

All I learnt is that the extension that triggered this is the JS Debugger Companion Extension, but that's where I got stuck.
